pageserver/src/metrics.rs

```diff
@@ -3034,13 +3034,104 @@ impl<F: Future<Output = Result<O, E>>, O, E> Future for MeasuredRemoteOp<F> {
 }
 
 pub mod tokio_epoll_uring {
-    use metrics::{register_int_counter, UIntGauge};
+    use std::{
+        collections::HashMap,
+        sync::{Arc, Mutex, Weak},
+    };
+
+    use metrics::{register_histogram, register_int_counter, Histogram, LocalHistogram, UIntGauge};
     use once_cell::sync::Lazy;
 
+    /// Shared storage for tokio-epoll-uring thread local metrics.
+    pub(crate) static THREAD_LOCAL_METRICS_STORAGE: Lazy<ThreadLocalMetricsStorage> =
+        Lazy::new(ThreadLocalMetricsStorage::new);
+
+    pub struct ThreadLocalMetricsStorage {
+        /// List of thread local metrics observers.
+        observers: Mutex<HashMap<u64, Weak<ThreadLocalMetrics>>>,
```

Comment:
Could also uses a `Vec`, and call `Vec::retain` when
1. register a system
2. drop `ThreadLocalStateInner`
3. `Collector::collect`

Using a `HashMap` eliminates the need for 3, and only need to insert/remove the relevant system.